controllers/lease_controller.go

```diff
@@ -0,0 +1,544 @@
+		env = append(env, getEnvVarFromSecrets("OCS_REGION", storeValues.SecretRef.Name, "region"))
+		env = append(env, getEnvVarFromSecrets("OCS_DISABLE_SSL", storeValues.SecretRef.Name, "disableSSL"))
+		env = append(env, getEnvVarFromSecrets("OCS_INSECURE_SKIP_VERIFY", storeValues.SecretRef.Name, "insecureSkipVerify"))
+	}
+
+	return env
+}
+
+func getEnvVarFromValues(name, value string) v1.EnvVar {
+	return v1.EnvVar{
+		Name:  name,
+		Value: value,
+	}
+}
+
+func getEnvVarFromFields(name, fieldPath string) v1.EnvVar {
+	return v1.EnvVar{
+		Name: name,
+		ValueFrom: &v1.EnvVarSource{
+			FieldRef: &v1.ObjectFieldSelector{
+				FieldPath: fieldPath,
+			},
+		},
+	}
+}
+
+func getEnvVarFromSecrets(name, secretName, secretKey string) v1.EnvVar {
+	return v1.EnvVar{
+		Name: name,
+		ValueFrom: &v1.EnvVarSource{
+			SecretKeyRef: &v1.SecretKeySelector{
+				LocalObjectReference: v1.LocalObjectReference{
+					Name: secretName,
+				},
+				Key: secretKey,
+			},
+		},
+	}
+}
+
+func getCompactJobCommands(etcd *druidv1alpha1.Etcd) []string {
+	command := []string{"" + "etcdbrctl"}
+	command = append(command, "compact")
+	command = append(command, "--data-dir=/var/etcd/data")
+	command = append(command, "--snapstore-temp-directory=/var/etcd/data/tmp")
+	command = append(command, "--enable-snapshot-lease-renewal=true")
+	command = append(command, "--full-snapshot-lease-name="+getFullLease(etcd))
+	command = append(command, "--delta-snapshot-lease-name="+getDeltaLease(etcd))
+
+	var quota int64 = 8 * 1024 * 1024 * 1024 // 8Gi
+	if etcd.Spec.Etcd.Quota != nil {
+		quota = etcd.Spec.Etcd.Quota.Value()
+	}
+	command = append(command, "--embedded-etcd-quota-bytes="+fmt.Sprint(quota))
+
+	if etcd.Spec.Etcd.EtcdDefragTimeout != nil {
+		command = append(command, "--etcd-defrag-timeout="+etcd.Spec.Etcd.EtcdDefragTimeout.Duration.String())
+	}
+
+	backupValues := etcd.Spec.Backup
+	if backupValues.EtcdSnapshotTimeout != nil {
+		command = append(command, "--etcd-snapshot-timeout="+backupValues.EtcdSnapshotTimeout.Duration.String())
+	}
+	storeValues := etcd.Spec.Backup.Store
+	if storeValues != nil {
+		provider, err := utils.StorageProviderFromInfraProvider(etcd.Spec.Backup.Store.Provider)
+		if err == nil {
+			command = append(command, "--storage-provider="+provider)
+		}
+
+		if storeValues.Prefix != "" {
+			command = append(command, "--store-prefix="+storeValues.Prefix)
+		}
+
+		if storeValues.Container != nil {
+			command = append(command, "--store-container="+*(storeValues.Container))
+		}
+	}
+
+	return command
+}
+
+// SetupWithManager sets up manager with a new controller and ec as the reconcile.Reconciler
+func (lc *LeaseController) SetupWithManager(mgr ctrl.Manager, workers int) error {
+	builder := ctrl.NewControllerManagedBy(mgr).WithOptions(controller.Options{
+		MaxConcurrentReconciles: workers,
+	})
+
+	return builder.
+		For(&druidv1alpha1.Etcd{}).
```

Comment:
This will result in reconciliations when there are updates (spec or status) to the `etcd` resource as well. IIUC, we don't need to reconcile at all if there are updates to `etcd` but only react to events from `lease`s, right? Maybe we can add a predicate here to skip enqueuing.

Reply:
Should the predicate process creation, deletion events for leases? WDYT? I think no. Because the leases are created with values "0" in holderIdentity always

Reply:
`Create` events happen when the controller is started for all existing leases. Thus, I think it's a good idea that we react to `Create` events.

Reply:
Done